### DISCUSSION TITLE
How does refraction in Mitsuba work?

### DISCUSSION THREAD

**Discussion by h-OUS-e** [mitsuba-renderer/mitsuba3#526] [Q&A] [answered]
Hello,

I would like to formulate a problem that involves lenses and refraction with mitsuba 3, but I just wanted to ask how is a ray refracted via a surface, vs a solid in Mitsuba3?

If I have two glass surfaces, like B, would I expect the Ray to pass out from the second surface as shown in the diagram? And would a solid glass object treat the entry and exit surfaces individually such that the position of the intersection of the Ray and top surface of the glass solid is different from the position of the intersection of the ray and the bottom surface of the glass solid?

![image](https://user-images.githubusercontent.com/66432003/217432738-f76e818c-e4e8-4c5a-b5af-a9df2d9e5d35.jpg)

**Comment by njroussel** [comment, marked as answer]
Hi @h-OUS-e 

You'll want to take a look at documentation of the [dielectric plugin](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_bsdfs.html#smooth-dielectric-material-dielectric) (and [thindielectric plugin ](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_bsdfs.html#thin-dielectric-material-thindielectric) for the figures).
All 4 options (is there a difference between C and D) in your diagram are possible by just assigning the `dielectric` BSDF to your `Shape`.

**Reply by h-OUS-e** [reply]
Awesome thank you! A followup question, why do we use dielectric in the caustic optimization tutorial instead of thindielectric given the Lens we are optimizing is a thin surface? Or would it count as one full solid since the vertices of the lens and the glass slab are touching?

**Reply by njroussel** [reply]
Imagine what would happen if the slab was a `thindielectric`. The outgoing rays would have the same direction as the incoming rays, thereby making it impossible to focus the the light such that it recreates the target image.

**Reply by h-OUS-e** [reply]
I see, I wasn’t sure the thindielectric would do that. But then, we did we specify where I’d the interior and exterior of the glass slab and the lens? How does mitsuba know that if they are two separate objects?

**Reply by njroussel** [reply]
I'm not quite sure I understand your questions. What are you referring to as glass slab and lens? There is only one object in the scene.

**Reply by h-OUS-e** [reply]
![1D8CB19F-7B28-4C2F-9EFD-0B5CE3160C6B](https://user-images.githubusercontent.com/66432003/217848329-27fe3cc2-303f-4e93-8055-f7fc4a336f6a.jpeg)

In the tutorial, the glass slab is divided in the glass rectangle with an id:lens (which will be optimized) and an open mesh with an id:slab. They’re two separate surfaces and not solids, so I was wondering if mitsuba automatically knows the interior and exterior of two surfaces sharing the same vertices and dielectric materiality.

**Reply by njroussel** [reply]
Ohhh I see! I had forgotten how this was setup in the tutorial.

Mitsuba doesn't track IOR changes with the ray, rather it computes it locally whenever there is an interaction. As to how the int/ext-erior are defined at each interaction, this is specified in the `dielectric` [plugin documentation](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_bsdfs.html#smooth-dielectric-material-dielectric): 
_Exterior and interior IOR values can be specified independently, where “exterior” refers to the side that contains the surface normal_

The tutorial takes a few shortcuts, but actually you would want to have a single watertight shape and during the optimization only apply changes to the vertices on the lens side of the object.

**Reply by h-OUS-e** [reply]
I see thank you!
